doc/design/migration_to_python.md

```diff
@@ -0,0 +1,479 @@
+# Migrating to Python
+
+The design is about migrating SQLFlow golang code that's related to ML to python.
+
+Language migration is difficult in common sense. The section [Flaws In The Current Architecture](#flaws-in-the-current-architecture) explains why the migration should happen. If you're already convinced that the migration is necessary, this section can be safely skipped and look over at any time.
+
+The section [A Quick View Of The Current Architecture](#a-quick-view-of-the-current-architecture) shows the distribution of golang  *ML code* in the current architecture. That's what to be migrated to python.
+
+The section [The Proposed Architecture](#the-proposed-architecture) illustrates an outline design about how the migration will be done.
+
+The last two sections explain how the proposed architecture solves the problem of the current architecture, as well as the preparations to be noted.
+
+## Flaws In The Current Architecture
+
+For SQLFlow, migrating to python might mean rewriting lots of existing golang code in python. Why should we migrate the golang *ML code* to python? Is is necessary? Is it the right time?
+
+After demonstrating the flaws of the current architecture in the following scenarios, the answer will be obvious.
+
+### Scenario 1: An ML Expert Wants A New Column Type
+
+Feature engineering is a very important part of machine learning. SQLFlow supports feature engineering via the `COLUMN` clause. The current `COLUMN` clause of SQLFlow is based on `tf.feature_column` and works only for TensorFlow. In the next version, the SQLFlow `COLUMN` clause will support more comprehensive feature columns for TensorFlow, PyTorch and XGBoost. 
+
+Suppose the user wants a `image_embedding_column ` that converts a input image into embedding for a recommender system. Adding such a new column in the current architecture requires the following main steps:
```

Comment:
> If we do not generate a python script, the workflow script is implemented by calling python platform.train/predict/evaluate functions? So golang should call python (rpc or native calls)? If so, how about make it more clear if there is any cross-language call?

1. golang starts a python process that calls the python function `execute()`
2. `execute()` generates a `.yaml` and submits the `.yaml` to Argo 
3. Each step in the `.yaml` starts the `step` binary
4. The `step` binary calls the python function `step()`
5. The `step()` python function calls python platform.train/predict/evaluate functions respectively.

Reply:
Does each step execute a SQL statement?

- Execute `step -e "SELECT ... TO TRAIN"` in each step:
    - the step image must have a step binary, java parser, generate python code, and call the python code in a sub-process
- Directly execute `sqlflow.train()` in each step:
    - the step image only contain `sqlflow` Python API for use, don't need to parse the statement again.

FYI.

Reply:
Yes, actually, the python function `sqlflow.execute step` is designed for the purpose, eventually, we'll call `python -m sqlflow.execute step` to execute each step. I propose to do this in the second phase rather than in the MVP phase, because this requires to reimplement code in the golang package `step` and `tablewriter`, which will distract us from the main purpose of the migration.

In the MVP phase, we still reuse the `step` binary and focus on the migration of the *ML code* in golang. In the 2nd phase of the migration, we can migrate the golang package `step` at a right time.